lib/native/source/libboolector/interface_wrap.c

```diff
@@ -3654,137 +3652,157 @@ SWIGEXPORT jobjectArray JNICALL Java_org_sosy_1lab_java_1smt_solvers_boolector_B
+  FILE *file = 0;
+  char *buffer = NULL;
+  int fileDesrc = -1;
+  long fileLength = 0;
   BoolectorNode *arg2 = (BoolectorNode *) 0 ;
-  char *tfnwd = addTemppathToFilename(filename);
+  char *tempfileName = addTemppathToFilename(filenameTemplate);
 
-  fd = mkstemp(tfnwd);
-  if(fd == -1) {
-    free(tfnwd);
-    perror("ERROR CREATING TEMPORARY FILE FOR");
+  if (tempfileName == NULL) {
+    perror("ERROR CREATING TEMPORARY FILE FOR BOOLECTOR_HELP_DUMP_NODE_SMT2");
+    SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "FileName for boolector_help_dump_node_smt2 may not be NULL");
+    return 0;
+  }
+
+  fileDesrc = mkstemp(tempfileName);
+  if (fileDesrc == -1) {
+    free(tempfileName);
+    perror("ERROR CREATING TEMPORARY FILE FOR BOOLECTOR_HELP_DUMP_NODE_SMT2");
     SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "FileDescriptor for boolector_help_dump_node_smt2 may not be NULL");
     return 0;
   }
-    
+
   (void)jenv;
   (void)jcls;
   arg1 = *(Btor **)&jarg1;
-  arg2 = *(BoolectorNode **)&jarg2; 
+  arg2 = *(BoolectorNode **)&jarg2;
 
-  f = fdopen(fd, "w+");
-  if(f==NULL) {
-    free(tfnwd);
-    unlink(tfnwd);
-    perror("ERROR: COULDNT DUMP NODE BECAUSE IT COULDNT CREATE A DUMP FILE"); 
+  file = fdopen(fileDesrc, "w+");
+  unlink(tempfileName);
+  if (file == NULL) {
+    perror("ERROR: COULDNT DUMP NODE BECAUSE IT COULDNT CREATE A DUMP FILE");
     SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "File for boolector_help_dump_node_smt2 may not be NULL");
     return 0;
   }
-    
+
   //write
-  boolector_dump_smt2_node(arg1, f, arg2);
-    
+  boolector_dump_smt2_node(arg1, file, arg2);
+  rewind(file);  //Just to be sure
+
   //read
-  if (f) {
-   fseek (f, 0, SEEK_END);
-    length = ftell (f);
-    fseek (f, 0, SEEK_SET);
-    buffer = malloc (length);
-    if (buffer) {
-      if(fread (buffer, 1, length, f) != (unsigned long)length) {
-        unlink(tfnwd); 
-        free(tfnwd);
-        free(buffer);
-        perror("ERROR READING FILE INTO BUFFER");
-        SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "Buffer for boolector_help_dump_node_smt2 may not be NULL");
-        return 0;
-      }
-    }
+  if (!file) {
+    perror("ERROR: FILE RETURNED BY BOOLECTOR_DUMP_SMT2_NODE IS NULL");
+    SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "File returned by boolector_dump_smt2_node() is NULL. boolector_help_dump_node_smt2 aborted.");
+    return 0;
   }
 
-  if (buffer) {
-    result = buffer;
+  if (fseek(file, 0, SEEK_END) != 0) {
+    perror("ERROR SEEKING FILE LENGTH");
+    SWIG_JavaThrowException(jenv, SWIG_JavaNullPointerException, "boolector_help_dump_node_smt2 could not determine the end of the used file");
+    return 0;
   }
-    
-  if (result) jresult = (*jenv)->NewStringUTF(jenv, (const char *)result);
-    
-  unlink(tfnwd); 
-  free(tfnwd);
-  fclose(f); 
+
+  fileLength = ftell(file);
+  rewind(file);
```

Comment:
Is rewind really needed here?

Reply:
I think so, yes.
The stream is set to the end of the file before calling `ftell()` to read the length. After that we should put it back to the beginning for `fread()` later.

Reply:
Is the old file descriptor not pointing to the start of the file any more? I have never though about it. Is a C-based file a real `stram`? Then this is ok.

Reply:
In C a file is a stream, yes.
The Object of type `FILE` has all the necassary information to control that stream.

Reply:
Ok, I see.
Would it be nicer to use `if ( fseek(file, 0L, SEEK_SET) != 0 ) {` instead of `rewind(file)`?
This would allow some error handling.